File: frontend/lib/api/campagne_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_config.dart';
import 'jwt_service.dart';
import '../models/campagne.dart';

/// Récupère toutes les campagnes (par utilisateur ou par projet).
Future<List<Campagne>> getCampagnes({int? utilisateurId, int? projetId}) async {
  final headers = await authHeaders();
  String url = '${ApiConfig.apiBaseUrl}/campagnes';
  if (projetId != null) {
    url = '$url?projetId=$projetId';
  } else if (utilisateurId != null) {
    url = '$url?utilisateurId=$utilisateurId';
  }
  final response = await http
      .get(Uri.parse(url), headers: headers)
      .timeout(ApiConfig.timeout);

  if (response.statusCode == 200) {
    final List<dynamic> jsonList = jsonDecode(response.body);
    return jsonList
        .map((json) => Campagne.fromJson(json as Map<String, dynamic>))
        .toList();
  }
  throw Exception('Erreur ${response.statusCode}: ${response.body}');
}

/// Récupère une campagne par son ID.
Future<Campagne> getCampagneById(int id) async {
  final headers = await authHeaders();
  final response = await http
      .get(Uri.parse('${ApiConfig.apiBaseUrl}/campagnes/$id'), headers: headers)
      .timeout(ApiConfig.timeout);

  if (response.statusCode == 200) {
    return Campagne.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
  }
  throw Exception('Erreur ${response.statusCode}: ${response.body}');
}

/// Crée une nouvelle campagne.
Future<Campagne> createCampagne({
  required String nom,
  String? description,
  String? zone,
  required int createurId,
  int? projetId,
}) async {
  final headers = await authHeaders();
  final response = await http
      .post(
        Uri.parse('${ApiConfig.apiBaseUrl}/campagnes'),
        headers: headers,
        body: jsonEncode({
          'nom': nom,
          'description': description,
          'zone': zone,
          'createurId': createurId,
          'projetId': projetId,
        }),
      )
      .timeout(ApiConfig.timeout);

  if (response.statusCode == 201) {
    return Campagne.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
  }
  throw Exception('Erreur ${response.statusCode}: ${response.body}');
}

/// Met à jour une campagne existante (PUT — tous les champs requis).
Future<Campagne> updateCampagne({
  required int id,
  required String nom,
  String? description,
  String? zone,
  bool? estCloturee,
  required int createurId,
}) async {
  final headers = await authHeaders();
  final response = await http
      .put(
        Uri.parse('${ApiConfig.apiBaseUrl}/campagnes/$id'),
        headers: headers,
        body: jsonEncode({
          'nom': nom,
          'description': description,
          'zone': zone,
          'estCloturee': estCloturee,
          'createurId': createurId,
        }),
      )
      .timeout(ApiConfig.timeout);

  if (response.statusCode == 200) {
    return Campagne.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
  }
  throw Exception('Erreur ${response.statusCode}: ${response.body}');
}

/// Met à jour partiellement une campagne (PATCH — seuls les champs fournis sont modifiés).
Future<Campagne> patchCampagne({
  required int id,
  String? nom,
  String? description,
  String? zone,
  bool? estCloturee,
}) async {
  final headers = await authHeaders();
  final body = <String, dynamic>{};
  if (nom != null) body['nom'] = nom;
  if (description != null) body['description'] = description;
  if (zone != null) body['zone'] = zone;
  if (estCloturee != null) body['estCloturee'] = estCloturee;

  final response = await http
      .patch(
        Uri.parse('${ApiConfig.apiBaseUrl}/campagnes/$id'),
        headers: headers,
        body: jsonEncode(body),
      )
      .timeout(ApiConfig.timeout);

  if (response.statusCode == 200) {
    return Campagne.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
  }
  throw Exception('Erreur ${response.statusCode}: ${response.body}');
}

/// Supprime une campagne.
Future<void> deleteCampagne(int id) async {
  final headers = await authHeaders();
  final response = await http
      .delete(
        Uri.parse('${ApiConfig.apiBaseUrl}/campagnes/$id'),
        headers: headers,
      )
      .timeout(ApiConfig.timeout);

  if (response.statusCode != 204) {
    throw Exception('Erreur ${response.statusCode}: ${response.body}');
  }
}

File: web/src/api/campagneApi.ts
import { request } from './request'
import type {
  CampagneRequestDto,
  CampagneResponseDto,
} from '../types'

/** Liste les campagnes d'un projet (le backend exige projetId). */
export async function getCampagnes(projetId: number): Promise<CampagneResponseDto[]> {
  return request<CampagneResponseDto[]>(`/campagnes?projetId=${projetId}`)
}

/** Récupère une campagne par ID. */
export async function getCampagneById(id: number): Promise<CampagneResponseDto> {
  return request<CampagneResponseDto>(`/campagnes/${id}`)
}

/** Crée une campagne. */
export async function createCampagne(payload: CampagneRequestDto): Promise<CampagneResponseDto> {
  return request<CampagneResponseDto>('/campagnes', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

/** Met à jour une campagne. */
export async function updateCampagne(id: number, payload: CampagneRequestDto): Promise<CampagneResponseDto> {
  return request<CampagneResponseDto>(`/campagnes/${id}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
}

/** Supprime une campagne. */
export async function deleteCampagne(id: number): Promise<void> {
  await request<unknown>(`/campagnes/${id}`, { method: 'DELETE' })
}

/** Met à jour partiellement une campagne (ex: clôturer / réouvrir). */
export async function patchCampagne(
  id: number,
  payload: { estCloturee: boolean },
): Promise<CampagneResponseDto> {
  return request<CampagneResponseDto>(`/campagnes/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  })
}
